File: app/src/main/java/com/example/nijimac103/itunestracker/service/viewModel/ArtistListViewModel.java
package com.example.nijimac103.itunestracker.service.viewModel;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;
import android.arch.lifecycle.LiveData;
import android.os.Handler;
import android.util.Log;

import com.example.nijimac103.itunestracker.service.model.ArtistList;
import com.example.nijimac103.itunestracker.service.repository.ArtistRepository;

import static com.example.nijimac103.itunestracker.service.util.CalcUtils.getRand;


public class ArtistListViewModel extends AndroidViewModel {

    public static final int DELAY_MINUTES = 60000;
    public static final int MAX_REQUEST_PER_MINUTE = 20;
    private int LIMIT = 30;
    public static final int ZERO = 0;
    public static final String MUSIC_VIDEO = "musicVideo";
    private LiveData<ArtistList> artistListObservable;
    private ArtistRepository repo;
    private int count;

    //TODO:決め打ちでなく、rankingからとってきてTrendとしてもいいかも
    private String[] defaultArtists = {"aliciakeys","ladygaga","caroleking","beatles","jamestaylor","ericclapton","beyonce","jamesbrown"};

    public ArtistListViewModel(Application application) {
        super(application);

        setUpRequestScheduler();

        repo = ArtistRepository.getInstance();

        artistListObservable = repo.getArtistList(defaultArtists[getRand(7)], MUSIC_VIDEO, LIMIT);
    }

    //LiveDataのゲッター
    public LiveData<ArtistList> getArtistListObservable() {

        return artistListObservable;
    }

    public void reloadArtists(CharSequence text) {
        if (count <= MAX_REQUEST_PER_MINUTE) {
            artistListObservable = repo.getArtistList(text.toString(), "musicVideo", LIMIT);
            count++;
        } else {
            Log.d("監視", count + "制限超過");
        }
    }

    private void setUpRequestScheduler() {

        final Handler handler = new Handler();

        handler.postDelayed(new Runnable() {
            @Override
            public void run() {
                // UIスレッド
                if (count > MAX_REQUEST_PER_MINUTE) {
                    Log.d("監視", count + "だから危険");
                } else {
                    Log.d("監視", count + "だから安全");
                }
                count = ZERO;
                handler.postDelayed(this, DELAY_MINUTES);
            }
        }, DELAY_MINUTES);
    }

}

File: app/src/main/java/jp/co/geisha/itunestracker/viewModel/ArtistListViewModel.kt
package jp.co.geisha.itunestracker.viewModel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import android.os.Handler
import android.text.TextUtils
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import jp.co.geisha.itunestracker.*
import jp.co.geisha.itunestracker.repository.ArtistRepository
import timber.log.Timber
import jp.co.geisha.itunestracker.ConstArrays.DEFAULT_ARTIST_LIST
import jp.co.geisha.itunestracker.api.entity.Artist
import jp.co.geisha.itunestracker.util.CalcUtils.getRand
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class ArtistListViewModel(application: Application) : AndroidViewModel(application) {

    sealed class Status {
        object Success : Status()
        object Loading : Status()
        data class Error(val message: String) : Status()
    }

    val videoListLiveData: MutableLiveData<Artist.Result> = MutableLiveData()
    val state: MutableLiveData<Status> = MutableLiveData()

    private val repository: ArtistRepository = ArtistRepository.getInstance()

    private val handler = Handler()
    private var count: Int = 0

    fun loadArtists(textInput: CharSequence) {
        viewModelScope.launch {
            state.value = Status.Loading
            if (count <= MAX_REQUEST_PER_MINUTE) {
                val searchText = if (TextUtils.isEmpty(textInput)) {
                    DEFAULT_ARTIST_LIST[getRand(DEFAULT_ARTIST_LIST.size)]
                } else {
                    textInput
                }
                getVideoListWithQuery(searchText.toString())
            } else {
                state.postValue(Status.Error("Possible limit exceeded , Now count == $count"))
            }
        }
    }

    private suspend fun getVideoListWithQuery(query: String) = withContext(Dispatchers.Default) {
        val result = repository.getMusicVideoList( query, MUSIC_VIDEO, LIMIT)
        if (result.isSuccessful) {
            state.postValue(Status.Success)
            videoListLiveData.postValue(result.body())
            count++
        } else {
            state.postValue(Status.Error("api call failed"))
        }
    }

    private val scheduler = object : Runnable {
        override fun run() {
            if (count > MAX_REQUEST_PER_MINUTE) {
                Timber.d("%s danger count", count)
            } else
                Timber.d("%s count safely", count)
            count = ZERO
            handler.postDelayed(this, DELAY_MINUTES.toLong())
        }
    }

    fun setUpRequestScheduler() {
        viewModelScope.launch {
            handler.postDelayed(scheduler, DELAY_MINUTES.toLong())
        }
    }

    override fun onCleared() {
        super.onCleared()
        handler.removeCallbacks {
            scheduler
        }
    }
}
